junit-platform-launcher/src/main/java/org/junit/platform/launcher/LauncherDiscoveryListener.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.launcher;
+
+import java.util.Optional;
+
+import org.junit.platform.engine.EngineDiscoveryListener;
+import org.junit.platform.engine.UniqueId;
+
+public interface LauncherDiscoveryListener extends EngineDiscoveryListener {
+
+	default void engineDiscoveryStarted(UniqueId engineId) {
+	}
+
+	default void engineDiscoveryFinished(UniqueId engineId, Optional<Throwable> failure) {
```

Comment:
No `Optional` in parameter types.

Either allow `null` as an indicator for no failure or provide two finished callbacks:
- `default void engineDiscoveryFinished(UniqueId engineId) {}`
- `default void engineDiscoveryFinishedWithFailure(UniqueId engineId, Throwable failure) {}`

Or wrap it in a `EngineDiscoveryFinishedResult` holder that has an `Optional<Throwable> failure()` accessor.

Reply:
> No `Optional` in parameter types.

We don't always follow that "rule" here. Nevertheless, you have a point and I think it would be more consistent to use a result object.